packages/labs/ssr/src/lib/declarative-style-dedupe.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {digestForTemplateResult} from '@lit-labs/ssr-client';
+import {CSSResult, CSSResultOrNative, TemplateResult} from 'lit';
+
+/**
+ * StyleDedupe is a utility that can be optionally passed to your
+ * SSR
+ */
+export class DeclarativeStyleDedupeUtility {
+  seenStyles = new Set<string>();
+  private hasEmittedDeclaration = false;
+
+  emitCustomElementDeclaration(): string {
+    if (this.hasEmittedDeclaration) {
+      return '';
+    }
+    this.hasEmittedDeclaration = true;
+    return `
+<script>
+
+class StyleModule extends HTMLElement {
+    static styleCache = new Map();
+
+    connectedCallback() {
+        // This attribute tells us that we expect the child styles
+        // to be cached.
+        if (this.hasAttribute("need-cached-styles")) {
+            return this.parsedCallback();
+        }
+
+        const obs = new MutationObserver(() => {
+            obs.disconnect();
+            this.parsedCallback();
+        });
+        obs.observe(this.parentNode, {childList: true, subtree: true});
```

Comment:
to definitively determine a tag close you can inject a `<script>` tag after it that does something like `document.currentScript.previousSibling.loadStyles()`

Reply:
Awesome! This is super helpful. To cache the first `style` element, we need the caching to happen after the `style-module` closing tag has been parsed (thus the `style` element exists within it).

I didn't realize `connectedCallback` ran at **tag open** time and was quite perplexed when there was no `style` element anywhere.

Reply:
After reading https://github.com/lit/lit/pull/4378#discussion_r1387367228 I can see an approach where no parse timing needs to be managed.
Still, this was a great comment to learn from. Thank you!

Reply:
yeah, you don't need to cache the style text upon the first style module element, just cache the element. But you do need to know when a style-module has no children so you can _fetch_ the style text. Knowing a tag will have no more children is.... hard.

But you actually don't need a separate script tag in this case - you could just use a different attribute. I believe all browsers will have parsed attributes by the time connectedCallback is called, so you can read a `load-styles="id"` attribute and then just go get the styles right then. I'm sure that'll be faster than having possibly 100s of script tags on the page.